core-addon/IonCore.js

```diff
@@ -14,11 +17,135 @@ module.exports = class IonCore {
       }
       this._openControlPanel();
     });
+
+    // Listen for messages from the options page.
+    browser.runtime.onMessage.addListener(
+      (m, s) => this._handleMessage(m, s));
   }
 
   _openControlPanel() {
     browser.runtime.openOptionsPage().catch(e => {
       console.error(`IonCore.js - Unable to open the control panel`, e);
     });
   }
+
+  /**
+   * Handles messages coming in from the options page.
+   *
+   * @param {Object} message
+   *        The payload of the message.
+   * @param {runtime.MessageSender} sender
+   *        An object containing informations about who sent
+   *        the message.
+   * @returns {Promise} The response to the received message.
+   *          It can be resolved with a value that is sent to the
+   *          `sender`.
+   */
+  _handleMessage(message, sender) {
+    console.log(`IonCore - received message ${message} from ${sender}`);
+
+    // We only expect messages coming from the embedded options page
+    // at this time. Discard anything else and report an error.
+    if (sender.url != browser.runtime.getURL(ION_OPTIONS_PAGE_PATH)) {
+      console.error(`IonCore - received message from unexpected sender`);
+      return Promise.reject();
+    }
+
+    switch (message.type) {
+      case "enrollment": {
+        // Let's not forget to respond `true` to the sender: the UI
+        // is expecting it.
+        return this._enroll().then(r => true);
+      } break;
+      default:
+        console.error(`IonCore - unexpected message type ${message.type}`);
+        return Promise.reject();
+    }
+  }
+
+  /**
+   * Enroll in the Ion platform.
+   *
+   * This sets up all the required information (e.g. Ion ID)
+   * and sets the relevant data to the pipeline.
+   *
+   * @returns {Promise} A promise resolved when the enrollment
+   *          is complete (does not block on data upload).
+   */
+  async _enroll() {
+    // TODO: Generate a proper random UUID.
```

Comment:
@rhelmer I ended up adding this as a safe placeholder to have the submission work. One thing we could do is just land this as it is, and follow up on proper UUID generation if you don't get to it :)

Reply:
I think it's OK to land as-is, I am going to attempt to use `browserify` so we could use https://github.com/uuidjs/uuid or something similar, that says it is compliant with the RFC. It's tempting to write our own but there are a few odd edge cases I'd like to avoid dealing with :)

Reply:
Until we get this figured out, you could add an experiments API that uses the Firefox UUID generator if you want? I took a look and it's implemented in C++ so probably not something we can pluck out easily ;)

Reply:
> 
> 
> Until we get this figured out, you could add an experiments API that uses the Firefox UUID generator if you want? I took a look and it's implemented in C++ so probably not something we can pluck out easily ;)

Great idea! Let me do this!